### PR TITLE
Feature/ui new log viewer partially working slider

### PR DESCRIPTION
**Pull request by PGuimarais**
- Introduces LogViewer's d3 timeline chart
- Incorporates design changes

### CODE REVIEW

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 1]
```diff
@@ -24,6 +24,40 @@ function LogViewerController ($scope) {
...
+  }
+  if(cols['level']){
+    theColumns.push('level');
+  }
+  if(cols['time']){
+    theColumns.push('time');
+  }
+
+  var collapseCount = 0;
+  this.collapseColumns = function() {
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: what is this being used for?

File: cdap-ui/app/directives/log-viewer/log-viewer.less [hunk 2]
```diff
@@ -270,25 +326,28 @@ my-log-viewer {
...
+          left: 20px;
         }
 
         .stack-trace {
           position: relative;
-          padding-bottom: 20px;
-          bottom: 10px;
-          height: 100%;
+          padding-bottom: 11px;
+          height: 101%;
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.less]: I don't think we should set height to 101%. If this height needs to be >100%, then there's something wrong somewhere else

File: cdap-ui/app/directives/timeline/timeline.less [hunk 3]
```diff
@@ -51,13 +80,23 @@ my-timeline {
         }
 
         .yellow-circle {
-          fill: yellow;
-          stroke: yellow;
+          fill: #f8b154;
+          stroke: #f8b154;
```
Comment (elfenheart) [inline cdap-ui/app/directives/timeline/timeline.less]: set these colors as variables

File: cdap-ui/app/directives/log-viewer/log-viewer.js (hunk 1 above)
Comment (PGuimarais) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: Collapsing the columns when pressing the arrow button on the 'Messages' header. They collapse one after the other, from right to left

Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: Is that according to design? I believe the design is to show message spanning the entire column. @bdmogal confirm?

Comment (bdmogal) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: Expand/collapse is only for the message column. it should have the table only contain the messages column, and reverse that behavior

**Comment by ajainarayanan** [comment]
screen shot 2016-07-07 at 1 09 39 pm

Timeline seems have some issue. The time looks odd.

**Comment by ajainarayanan** [comment]
@PGuimarais We don't dispatch any action once we choose a startime on the timeline do we? Seems like we added the store and TimelineController but we didn't stitch them up together to communicate the start time in the timeline.

File: cdap-ui/app/directives/timeline/timeline-ctrl.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+function TimelineController ($scope, LogViewerStore) {
+  console.log('From TimelineController: ' + LogViewerStore.getState().startTime);
+
+  this.testThis = 'hello';
+
+  LogViewerStore.subscribe(() => {
+    this.testTime = LogViewerStore.getState().startTime;
+  });
+
+  // LogViewerStore.subscribe(() => {
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/timeline/timeline-ctrl.js]: We should probably remove comments. we can always go back in history and get them.

File: cdap-ui/app/directives/timeline/timeline-ctrl.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  //   this.testScroll = LogViewerStore.getState();
+  // });
+
+  this.updateTheSlider = function() {
+    LogViewerStore.dispatch({
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/timeline/timeline-ctrl.js:20]: I don't think this function is used. In the link function, when we change the slider we need to call this function to update the store and the subscribers. This is again part of stitching things together, the view - store & the controller.

File: cdap-ui/app/directives/timeline/timeline-ctrl.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+      type: 'START_TIME',
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/timeline/timeline-ctrl.js]: This should use the `LOGVIEWERSTORE_ACTIONS`.

File: cdap-ui/app/directives/timeline/timeline-ctrl.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+      payload: {
+        startTime: Date.now()
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/timeline/timeline-ctrl.js]: I am guessing when we fix the function we will fix this too. The startTime should be the time selected using the slider.

File: cdap-ui/app/directives/timeline/timeline-ctrl.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+  LogViewerStore.subscribe(() => {
+    this.testTime = LogViewerStore.getState().startTime;
+  });
+
+  // LogViewerStore.subscribe(() => {
+  //   this.testScroll = LogViewerStore.getState();
+  // });
+
+  this.updateTheSlider = function() {
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/timeline/timeline-ctrl.js]: I am not sure about the name. May be `updateStartTimeInStore` or something smaller?

File: cdap-ui/app/directives/timeline/timeline.js [hunk 9]
```diff
@@ -115,78 +131,226 @@ function link (scope, element) {
...
+      sliderX = val;
+
+      if(sliderX > pinX){
+        updatePin(sliderX);
+      }
+
+      leftHandle.attr('x', val);
+      console.log('Left slider is at : ' + xScale.invert(val));
+
+      sliderBar.attr('d', 'M0,0V0H' + val + 'V0');
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/timeline/timeline.js:127]: This should eventually call `updateTheSlider` or appropriate function that updates the `LogViewerStore`.

File: cdap-ui/app/directives/log-viewer/log-viewer.html [hunk 10]
```diff
@@ -58,12 +58,12 @@
...
       <thead>
         <th class="time log-header"
-            ng-if="!LogViewer.isMessageExpanded && LogViewer.configOptions.time">
+            ng-if="!LogViewer.isMessageExpanded && LogViewer.configOptions.time && !LogViewer.hiddenColumns.time">
           Time
         </th>
 
         <th class="level log-header header-level"
-            ng-if="!LogViewer.isMessageExpanded && LogViewer.configOptions.level">
+            ng-if="!LogViewer.isMessageExpanded && LogViewer.configOptions.level && !LogViewer.hiddenColumns.level">
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.html]: if we are collapsing one by one, then `isMessageExpanded` is not needed

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 11]
```diff
@@ -14,16 +14,66 @@
...
+      }
+    } else {
+      collapseCount = 0;
+      for(var key in this.hiddenColumns){
+        if(this.hiddenColumns.hasOwnProperty(key)){
+          this.hiddenColumns[key] = false;
+        }
+      }
+    }
+  }.bind(this);
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: remove `bind(this)` and use `=>` ES6 function

**Comment by ajainarayanan** [comment]
The time in the log viewer tables seems to have some issues. 
screen shot 2016-07-07 at 9 54 45 pm

**Comment by ajainarayanan** [comment]
screen shot 2016-07-07 at 9 55 35 pm
Also when change the slider in the timeline directive, I don't see any dots in the timeline but I still see three logs in the logviewer table. Some issue with either time or with the log-viewer table?

File: cdap-ui/app/directives/timeline/timeline-ctrl.js [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+function TimelineController ($scope, LogViewerStore, LOGVIEWERSTORE_ACTIONS) {
+  console.log('From TimelineController: ' + LogViewerStore.getState().startTime);
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/timeline/timeline-ctrl.js]: We should probably remove `console.log` statements.

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 13]
```diff
@@ -14,16 +14,66 @@
...
   'ngInject';
 
+  LogViewerStore.subscribe(() => {
+    this.logStartTime = LogViewerStore.getState().startTime;
+    updateTheTable(this.logStartTime);
+  });
+
+  //Use this function to update what is rendered
+  function updateTheTable(value) {
+    console.log('From LogViewerController: ' + value);
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: Here too. We should probably remove `console.log` statements. Also if this function is not being used to update the table then we should probably remove the function too.

**Comment by ajainarayanan** [comment]
Follow up on [this comment](https://github.com/caskdata/cdap/pull/5983#issuecomment-231274810) I think we are using two different data for timeline and log viewer. We should be using the same data so the dots in the timeline reflects the the logs in the logviewer table.

**Comment by ajainarayanan** [comment]
One more thing I noticed is that when we click on a row that has no stack trace we open an empty gray container. We should probably only open up a row if it has a stack trace which I believe the design had some indication. We had this in the previous PR where we had a arrow on the message if it had a stack trace. Any specific reason why we removed it?

**Comment by elfenheart** [comment]
@ajainarayanan I don't think we can use the same data for timeline and log viewer. The timeline should be rendering the full logs timeline, but the table only rendering partial log. Let's say the run has finished and it ran for 10 minutes. The timeline should show the dots for the full 10 minutes, but the table initially will only show 50 logs line

**Comment by PGuimarais** [comment]
Currently keeping the incorrectly rendered date, as we are using sample data and will be using a standardized format once we integrate the backend

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 14]
```diff
@@ -136,6 +196,14 @@ function LogViewerController ($scope) {
     },
   ];
 
+  var displayData = this.data.slice();
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: why are you slicing this.data?

File: cdap-ui/app/directives/timeline/timeline.js [hunk 15]
```diff
@@ -16,177 +16,221 @@
...
+    xScale = d3.time.scale().range([0, (maxRange)]);
+    xScale.domain(d3.extent(timelineData, function(d) {
+      return d.time;
+    }));
+    //Define the axes and ticks
+    let xAxis = d3.svg.axis().scale(xScale)
+        .orient('bottom')
+        .innerTickSize(-40)
+        .tickPadding(7)
+        .ticks(8);
```
Comment (elfenheart) [inline cdap-ui/app/directives/timeline/timeline.js:55]: I'm pretty sure this doesn't work. Ticks on time.scale does not accept integer. 

On another note, I believe the design spec wants this to be 12 ticks if I remember my conversation with @bdmogal correctly

File: cdap-ui/app/directives/timeline/timeline.js [hunk 16]
```diff
@@ -16,177 +16,221 @@
...
-      .attr('transform', 'translate(0,' + (height - 10) + ')')
+      .attr('class', 'xaxis-bottom')
+      .attr('transform', 'translate(' + ( (paddingLeft + paddingRight) / 2) + ',' + (height - 20) + ')')
       .call(xAxis);
-  }
+
+    //attach handler to brush
+    brush = d3.svg.brush()
+        .x(xScale)
+        .extent([0,0])
```
Comment (elfenheart) [inline cdap-ui/app/directives/timeline/timeline.js]: is this `.extent`needed?

File: cdap-ui/app/directives/timeline/timeline.js [hunk 17]
```diff
@@ -16,177 +16,221 @@
...
+    //Create the 3 bars used to represent the slider
+    let sliderBar = svg.append('g')
+      .attr('class', 'slider leftSlider')
+      .call(d3.svg.axis()
+        .scale(xScale)
+        .tickSize(0)
+        .tickFormat(''))
+      .select('.domain')
+      .select(function() {
+        return this.parentNode.appendChild(this.cloneNode(true));
```
Comment (elfenheart) [inline cdap-ui/app/directives/timeline/timeline.js]: Can you add a comment on the code what this is?

File: cdap-ui/app/directives/timeline/timeline.js (hunk 15 above)
Comment (PGuimarais) [inline cdap-ui/app/directives/timeline/timeline.js:55]: @elfenheart - the ticks function actually does accept an integer. It is clear in the 4.0 documentation, but not so much in the 3.x. Spoke with @bdmogal and we will let it automatically scale for simplicity at the moment. The ticks produce what d3 sees as a desired interval of time ; with 8 as the parameter it produces 5 ticks given the sample dataset, with 10 set it produces 16. Let's fine tune this later.

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 18]
```diff
@@ -136,6 +196,15 @@ function LogViewerController ($scope) {
     },
   ];
 
+  var tableData = this.data;
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: setting this to a new variable is unnescessary

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 19]
```diff
@@ -136,6 +196,15 @@ function LogViewerController ($scope) {
...
+  tableData.forEach(function(element, index) {
+    let formattedDate = new Date(tableData[index].time);
+    tableData[index].time = formattedDate;
+    tableData[index].displayTime = ((formattedDate.getMonth() + 1) + '/' + formattedDate.getDate() + '/' + formattedDate.getFullYear() + ' ' + formattedDate.getHours() + ':' + formattedDate.getMinutes() + ':' + formattedDate.getSeconds());
+  });
+
+  //this.data = tableData;
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: remove

**Comment by elfenheart** [comment]
👍  LGTM

squash before merging

**Pull request merged**